tools/cbmc/proofs/HTTP/global_state_HTTP.c

```diff
@@ -123,103 +183,193 @@ int is_valid_NetworkInterface(IotNetworkInterface_t *netif) {
-    pConnectionHandle->pNetworkConnection = safeMalloc(sizeof(_connHandle_t));
-    pConnectionHandle->pNetworkInterface = newNetworkInterface();
+    // network connection just points to an allocated memory object
+    pConnectionHandle->pNetworkConnection = safeMalloc(1);
+    pConnectionHandle->pNetworkInterface = allocate_NetworkInterface();
+    // ???: Should reqQ initialization be an assumption?
     pConnectionHandle->reqQ.pPrevious = &(pConnectionHandle->reqQ);
     pConnectionHandle->reqQ.pNext = &(pConnectionHandle->reqQ);
+    // ???: Should respQ initialization be an assumption?
     pConnectionHandle->respQ.pPrevious = &(pConnectionHandle->respQ);
     pConnectionHandle->respQ.pNext = &(pConnectionHandle->respQ);
   }
   return pConnectionHandle;
 }
 
+int is_valid_IotConnectionHandle(IotHttpsConnectionHandle_t handle) {
+  return
+    handle->pNetworkConnection &&
+    handle->pNetworkInterface &&
+    is_valid_NetworkInterface(handle->pNetworkInterface);
+}
+
+/****************************************************************
+ * IotHttpsResponseHandle constructor
+ ****************************************************************/
+
 /* Creates a Response Handle and assigns memory accordingly. */
-IotHttpsResponseHandle_t newIotResponseHandle() {
+IotHttpsResponseHandle_t allocate_IotResponseHandle() {
   IotHttpsResponseHandle_t pResponseHandle = safeMalloc(sizeof(_resHandle_t));
   if(pResponseHandle) {
     uint32_t len;
-    pResponseHandle->pBody = malloc(len);
-    pResponseHandle->pHttpsConnection = newIotConnectionHandle();
-    pResponseHandle->pHeaders = ((_resHandle_t*)pResponseHandle)->data;
-    pResponseHandle->pHeadersCur = pResponseHandle->pHeaders;
-    pResponseHandle->pHeadersEnd = pResponseHandle->pHeaders + sizeof(((_resHandle_t*)pResponseHandle)->data);
-    pResponseHandle->httpParserInfo.readHeaderParser.data = pResponseHandle;
+
     pResponseHandle->httpParserInfo.parseFunc = http_parser_execute;
-    pResponseHandle->pReadHeaderValue = safeMalloc(sizeof(uint32_t));
-    __CPROVER_assume(pResponseHandle->readHeaderValueLength >= 0 &&
-    pResponseHandle->readHeaderValueLength <= (pResponseHandle->pHeadersEnd - pResponseHandle->pHeaders));
-    pResponseHandle->pReadHeaderValue = malloc(pResponseHandle->readHeaderValueLength);
+
+    pResponseHandle->pBody = safeMalloc(len);
+    pResponseHandle->pHttpsConnection = allocate_IotConnectionHandle();
+    pResponseHandle->pReadHeaderField =
+      safeMalloc(pResponseHandle->readHeaderFieldLength);
+    pResponseHandle->pReadHeaderValue =
+      safeMalloc(pResponseHandle->readHeaderValueLength);
   }
   return pResponseHandle;
 }
 
+int is_valid_IotResponseHandle(IotHttpsResponseHandle_t pResponseHandle) {
+  int required =
+    __CPROVER_same_object(pResponseHandle->pHeaders,
+			  pResponseHandle->pHeadersCur) &&
+    // ???: Does this overconstrain End?
+    __CPROVER_same_object(pResponseHandle->pHeaders,
+			  pResponseHandle->pHeadersEnd);
```

Comment:
Ok I looked at the bottom part and you are simply constraining pHeaders to be anywhere after pHeadersCur. So ignore my last comment.